File: weather-app/src/main/java/rcgonzalezf/org/weather/common/analytics/AnalyticsManager.java
package rcgonzalezf.org.weather.common.analytics;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import java.util.HashSet;
import java.util.Set;
import rcgonzalezf.org.weather.BuildConfig;
import rcgonzalezf.org.weather.utils.WeatherUtils;

public class AnalyticsManager {

  static final String MOBILE = "Mobile";
  static final String UNKNOWN = "Unknown";
  static final String NONE = "None";
  static final String WIFI = "Wifi";

  public static final String ANDROID_VERSION = "ANDROID_VERSION";
  public static final String APP_VERSION = "APP_VERSION";
  public static final String NETWORK = "NETWORK";
  public static final String MULTIPANE = "MULTIPANE";

  @VisibleForTesting
  static String sAndroidVersion = Build.VERSION.RELEASE;
  @VisibleForTesting static String sAppVersion = BuildConfig.VERSION_NAME;
  private Context mContext;
  private Set<AnalyticsObserver> mAnalyticsObservers = new HashSet<>();
  private AnalyticsBaseData mAnalyticsBaseData;
  private String mScreenName;

  public AnalyticsManager(Context context) {
    mContext = context;
    mAnalyticsBaseData = new AnalyticsBaseData();
    mAnalyticsBaseData.data().put(ANDROID_VERSION, sAndroidVersion);
    mAnalyticsBaseData.data().put(APP_VERSION, sAppVersion);
    mAnalyticsBaseData.data().put(MULTIPANE, Boolean.toString(isMultipane()));
  }

  public void addObserver(AnalyticsObserver analyticsObserver) {
    mAnalyticsObservers.add(analyticsObserver);
  }

  private boolean isMultipane() {
    return WeatherUtils.isXLargeTablet(mContext);
  }

  void notifyOnScreenLoad(String screenName) {
    mScreenName = screenName;
    addBaseData();

    for (AnalyticsObserver analyticsObserver : mAnalyticsObservers) {
      analyticsObserver.onScreen(screenName, mAnalyticsBaseData);
    }
  }

  @NonNull
  @NetworkType private String getNetworkType() {
    ConnectivityManager connectivityManager =
        (ConnectivityManager) mContext.getSystemService(Context.CONNECTIVITY_SERVICE);
    NetworkInfo activeNetwork = connectivityManager.getActiveNetworkInfo();
    String networkType;

    if (activeNetwork != null) {
      if (activeNetwork.getType() == ConnectivityManager.TYPE_WIFI) {
        networkType = WIFI;
      } else if (activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE) {
        networkType = MOBILE;
      } else {
        networkType = UNKNOWN;
      }
    } else {
      networkType = NONE;
    }
    return networkType;
  }

  void removeObserver(AnalyticsObserver analyticsObserver) {
    mAnalyticsObservers.remove(analyticsObserver);
  }

  void notifyOnAction(AnalyticsEvent analyticsEvent) {
    addBaseData();
    for (AnalyticsObserver analyticsObserver : mAnalyticsObservers) {
      analyticsObserver.onAction(analyticsEvent, mScreenName, mAnalyticsBaseData);
    }
  }

  private void addBaseData() {
    mAnalyticsBaseData.data().put(NETWORK, getNetworkType());
  }
}

File: weather-app/src/main/java/rcgonzalezf/org/weather/common/analytics/AnalyticsManager.kt
package rcgonzalezf.org.weather.common.analytics

import android.content.Context
import android.net.ConnectivityManager
import android.os.Build
import androidx.annotation.VisibleForTesting
import rcgonzalezf.org.weather.BuildConfig
import rcgonzalezf.org.weather.utils.WeatherUtils.isXLargeTablet
import java.util.HashSet

open class AnalyticsManager(private val context: Context) {
    private val analyticsObservers: MutableSet<AnalyticsObserver> = HashSet()
    private val analyticsBaseData: AnalyticsBaseData = AnalyticsBaseData()
    private lateinit var screenName: String

    init {
        analyticsBaseData.data()[ANDROID_VERSION] = BUILD_ANDROID_VERSION
        analyticsBaseData.data()[APP_VERSION] = BUILD_APP_VERSION
        analyticsBaseData.data()[MULTIPANE] = java.lang.Boolean.toString(isMultipane)
    }

    companion object {
        const val MOBILE = "Mobile"
        const val UNKNOWN = "Unknown"
        const val NONE = "None"
        const val WIFI = "Wifi"
        const val ANDROID_VERSION = "ANDROID_VERSION"
        const val APP_VERSION = "APP_VERSION"
        const val NETWORK = "NETWORK"
        const val MULTIPANE = "MULTIPANE"

        @JvmField
        @VisibleForTesting
        var BUILD_ANDROID_VERSION = Build.VERSION.RELEASE

        @JvmField
        @VisibleForTesting
        var BUILD_APP_VERSION = BuildConfig.VERSION_NAME
    }

    fun addObserver(analyticsObserver: AnalyticsObserver) {
        analyticsObservers.add(analyticsObserver)
    }

    private val isMultipane: Boolean
        get() = isXLargeTablet(context)

    open fun notifyOnScreenLoad(screenName: String) {
        this.screenName = screenName
        addBaseData()
        for (analyticsObserver in analyticsObservers) {
            analyticsObserver.onScreen(screenName, analyticsBaseData)
        }
    }

    @get:NetworkType
    private val networkType: String
        get() {
            val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetwork = connectivityManager.activeNetworkInfo
            val networkType: String
            networkType = if (activeNetwork != null) {
                if (activeNetwork.type == ConnectivityManager.TYPE_WIFI) {
                    WIFI
                } else if (activeNetwork.type == ConnectivityManager.TYPE_MOBILE) {
                    MOBILE
                } else {
                    UNKNOWN
                }
            } else {
                NONE
            }
            return networkType
        }

    fun removeObserver(analyticsObserver: AnalyticsObserver?) {
        analyticsObservers.remove(analyticsObserver)
    }

    open fun notifyOnAction(analyticsEvent: AnalyticsEvent) {
        addBaseData()
        for (analyticsObserver in analyticsObservers) {
            analyticsObserver.onAction(analyticsEvent, screenName, analyticsBaseData)
        }
    }

    private fun addBaseData() {
        analyticsBaseData.data()[NETWORK] = networkType
    }
}
